appsrc/dcfile/dcencap.man
.TH DCENCAP 1 "19 April 2009" "DICOM PS3" "DICOM PS3 - Encapsulate in DICOM file"
.SH NAME
dcencap \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Encapsulate compressed stream in a DICOM file
.SH SYNOPSIS
.HP 10
.B dcencap
.B dicomheader rawpixeldata
.B \-transfersyntax " uid"
[
.B \-itemsize " nbytes"
|
.B \-singleitem
]
[
.B \-v|verbose|vv|veryverbose
]
.so man1/optin.so
.so man1/optin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B dcencap
reads the named dicom header input file and raw pixel data bitstream input file
and copies the information and pixel data into a new dicom file with the
pixel data encapsulated and the specified transfer syntax.
.LP
.SH OPTIONS
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1), though the input
filenames must be explicitly specified, ie. they cannot be redirected from standard input. The transfer syntax
must be explicitly specifed for the DICOM output file using the usual \-ts option, and it obviously has to
be one of the encapsulated transfer syntaxes. Options specific to this program are:
.TP
.B \-singleitem
.RS
Only a single encapsulated item is written, regardless of the number of frames (e.g., for MPEG).
.RE
.TP
.B \-itemsize " nbytes"
.RS
The number of bytes to include in each encapsulated item. Defaults to number of rows times number of columns times bytes per pixel in a single frame, unless it exceeds the size of what will fit in a 32 bit length field; obviously this calculation has no meaning if the data stream
is compressed.
.RE
.TP
.B \-verbose
.RS
Describe fragmentation process and number of bytes processed.
.RE
.TP
.B \-veryverbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
% dcencap -v -singleitem video_002.src video_002.mpg -of video_002.dcm -ts 1.2.840.10008.1.2.4.100
.RE
Using frameSize 304128
.RE
Using itemSize 136857600
.RE
Doing single item 
.RE
Wrote total of 9072492
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS
Cannot add further attributes after the encapsulated Pixel Data attribute.
.LP
Cannot separate concatenated multi-frame compressed data bitsream into separate fragments (e.g., to build a multi-frame JPEG encapsulated encoding).



CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcencap.cc
#include "attrmxls.h"
#include "mesgtext.h"
#include "getoptns.h"
#include "ioopt.h"
#include "dcopt.h"
#include "elmconst.h"
#include "attrval.h"

int
main(int argc, char *argv[])
{
	bool bad=false;

	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options1(options);
	InputOptions		input_options2(options);
	DicomOutputOptions 	dicom_output_options(options);

	unsigned itemSize=0;
	(void)options.get("itemsize",itemSize);
	if (itemSize%2 != 0) {
		cerr << EMsgDC("Item size must be even length") << endl;
		bad=true;
	}
	
	bool singleitem = options.get("singleitem");
	if (singleitem) {
		if (itemSize != 0) {
			cerr << EMsgDC("Cannot specify -itemsize if -singleitem") << endl;
			bad=true;
		}
	}
	
	bool veryverbose=options.get("veryverbose") || options.get("vv");
	bool verbose=options.get("verbose") || options.get("v") || veryverbose;

	bool writeListPreceedingPixelData = true;
	bool writePixelDataAttributeBase = true;
	bool writePixelDataFragments = true;
	bool writePixelDataAttributeSequenceDelimiter = true;

	dicom_input_options1.done();
	input_options2.done();
	dicom_output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener1(
		options,dicom_input_options1.filename);
	InputOpenerFromOptions input_opener2(
		options,input_options2.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << dicom_input_options1.errors();
	cerr << input_options2.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener1.errors();
	cerr << input_opener2.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options1.good()
	 || !input_options2.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener1.good()
	 || !input_opener2.good()
	 || !output_opener.good()
	 || !options
	|| bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options1.usage()
			<< input_options2.usage()
			<< dicom_output_options.usage()
			<< " [-itemsize nbytes|-singleitem]"
			<< " [-v|-verbose|-vv|-veryverbose]"
			<< MMsgDC(InputFile) << " " << MMsgDC(InputFile)
			<< "[" << MMsgDC(OutputFile) << "]"
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din1(*(istream *)input_opener1,
		dicom_input_options1.transfersyntaxuid,
		dicom_input_options1.usemetaheader);
	istream in2(input_opener2);
	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream  log(cerr);
	if (veryverbose) log << "******** While reading DICOM dataset ... ********" << endl; 
	list.read(din1,&log,veryverbose,0xffffffff,true,dicom_input_options1.uselengthtoend,dicom_input_options1.ignoreoutofordertags);
	
	if (!list.good()) {
		log << list.errors()
		    << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}
	else {
		if (writeListPreceedingPixelData) {
			if (!usualManagedAttributeListWrite(list,dout,
				dicom_output_options,log,veryverbose)) success=false;
		}
		if (writePixelDataAttributeBase) {
			// now construct and write an undefined length PixelData attribute "manually"
		
			dout << Uint16(0x7fe0);
			dout << Uint16(0x0010);
			if (dout.getTransferSyntaxInUse()->isExplicitVR()) {
				dout << 'O';
				dout << 'B';
				dout << Uint16(0);
			}
			dout << Uint32(0xffffffffl);	// VL
		
			// don't forget the empty item offset table "item" ...
		
			dout << Uint16(0xfffe);
			dout << Uint16(0xe000);
			dout << Uint32(0);		// VL
		}
		if (writePixelDataFragments) {		
			Uint16 numberOfFrames=1;
			Uint32 frameSize=0;
			{
				Attribute *aNumberOfFrames = list[TagFromName(NumberOfFrames)];
				if (aNumberOfFrames) {
					numberOfFrames=AttributeValue(aNumberOfFrames);
				}
			
				Uint16 vRows = 0;
				Attribute *aRows = list[TagFromName(Rows)];
				if (!aRows) {
					log << WMsgDC(MissingAttribute)
					<< " - \"Rows\""
					<< endl;
				}
				else {
					vRows=AttributeValue(aRows);
				}
				Uint16 vColumns = 0;
				Attribute *aColumns = list[TagFromName(Columns)];
				if (!aColumns) {
					log << WMsgDC(MissingAttribute)
					<< " - \"Columns\""
					<< endl;
				}
				else {
					vColumns=AttributeValue(aColumns);
				}
				Uint16 vBitsAllocated = 0;
				Attribute *aBitsAllocated = list[TagFromName(BitsAllocated)];
				if (!aBitsAllocated) {
					log << WMsgDC(MissingAttribute)
					<< " - \"BitsAllocated\""
					<< endl;
				}
				else {
					vBitsAllocated=AttributeValue(aBitsAllocated);
				}
				Uint16 vSamplesPerPixel = 1;
				Attribute *aSamplesPerPixel = list[TagFromName(SamplesPerPixel)];
				if (!aSamplesPerPixel) {
					log << WMsgDC(MissingAttribute)
					<< " - \"SamplesPerPixel\""
					<< endl;
				}
				else {
					vSamplesPerPixel=AttributeValue(aSamplesPerPixel);
				}
				frameSize=((Uint32)vRows)*vColumns*vSamplesPerPixel*((vBitsAllocated-1)/8+1);
			}
			if (frameSize == 0) {
				log << WMsgDC("Cannot compute frame size from Image Pixel Module attributes") << endl;
				frameSize=0xFFFFFFFE;
			}
			{
				if (verbose) {
					log << "Using frameSize " << dec << frameSize << endl; 
				}
				if (singleitem) {
					itemSize=numberOfFrames*frameSize;
					if (itemSize%2 != 0) ++itemSize;
					if (itemSize > 0xFFFFFFFE) itemSize=0xFFFFFFFE;	// maximum length of defined length item
				}
				else {
					if (itemSize == 0) {
						itemSize=(unsigned)frameSize;
						if (itemSize%2 != 0) ++itemSize;
						if (itemSize > 0xFFFFFFFE) itemSize=0xFFFFFFFE;	// maximum length of defined length item
					}
				}
				if (verbose) {
					log << "Using itemSize " << dec << itemSize << endl; 
				}
				unsigned long total=0;
				char *buffer = new char[itemSize];
				Assert(buffer);
				if (singleitem) {
					if (verbose) {
						log << "Doing single item " << endl; 
					}
					in2.read(buffer,itemSize);
					int bytesRead=in2.gcount();
					if (bytesRead > 0) {
						int bytesWritten = bytesRead;
						if (bytesWritten%2 != 0) {
							if (verbose) {
								log << "Padding odd length fragment" << endl; 
							}
							++bytesWritten;
							Assert(bytesWritten <= itemSize);
							buffer[bytesRead]=(char)0xff;	// pad character is 0xff for JPEG etc, after EOI
						}
						// write Item
						dout << Uint16(0xfffe);
						dout << Uint16(0xe000);
						dout << Uint32(bytesWritten);		// VL
						dout.write(buffer,bytesWritten);
						total+=bytesWritten;
					}
				}
				else {
					unsigned f;
					for (f=0; f < numberOfFrames; ++f) {
						if (verbose) {
							log << "Doing frame " << dec << f << endl; 
						}
						Uint32 bytesLeftInFrame=frameSize; 
						while (bytesLeftInFrame > 0) {
							unsigned bytesThisItem = bytesLeftInFrame > itemSize ? itemSize : (unsigned)bytesLeftInFrame;
							if (in2.eof()) {
								// This is not an error if data was compressed; we just reached the end of it ...
								//log << EMsgDC("Premature EOF - giving up") << endl;
								//success=false;
								bytesLeftInFrame=0;
								f=numberOfFrames;
							}
							else {
								in2.read(buffer,bytesThisItem);
								int bytesRead=in2.gcount();
								if (verbose) {
									log << "Wanted " << dec << bytesThisItem << " read " << bytesRead << endl; 
								}
								if (bytesRead > 0) {
									int bytesWritten = bytesRead;
									if (bytesWritten%2 != 0) {
										if (verbose) {
											log << "Padding odd length fragment" << endl; 
										}
										++bytesWritten;
										Assert(bytesWritten <= itemSize);
										buffer[bytesRead]=(char)0xff;	// pad character is 0xff for JPEG etc, after EOI
									}
									// write Item
									dout << Uint16(0xfffe);
									dout << Uint16(0xe000);
									dout << Uint32(bytesWritten);		// VL
									dout.write(buffer,bytesWritten);
									bytesLeftInFrame-=bytesRead;
									Assert(bytesLeftInFrame == 0 || bytesWritten%2 == 0);	// i.e. didn't pad in middle of item
									total+=bytesWritten;
								}
								else {
									// This is not an error if data was compressed; we just reached the end of it ...
									//log << EMsgDC("Read returns zero length - giving up") << endl;
									//success=false;
									bytesLeftInFrame=0;
									f=numberOfFrames;
								}
							}
						}
					}
				}
				if (verbose) {
					log << "Wrote total of " << dec << total << endl; 
				}
			}
			if (writePixelDataAttributeSequenceDelimiter) {
				dout << Uint16(0xfffe);
				dout << Uint16(0xe0dd);
				dout << Uint32(0);	// VL
			}
			dout.flush();
		}
	}

	return success ? 0 : 1;
}

	
